### PR TITLE
add example task for dynamic foraging

### PR DESCRIPTION
**Pull request by hanhou**
- #10

### CODE REVIEW

File: examples/task_dynamic_foraging.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    
+    
+    ex_task = DynamicForagingTask(
+        version="1.0.0",
+        description="Phase B in Han's slides (block = [10, 20, 5], p_sum = 0.8, p_ratio = [1:0])",
+        task_parameters=ex_parameters
+        )
+    print(ex_task)
+
+    # Update Task parameters individually
```
Comment (jwong-nd) [inline examples/task_dynamic_foraging.py:182]: Please remove lines 182-191, these lines do not do anything, and actually add non-schema fields to task parameters!

**Review by jwong-nd — CHANGES_REQUESTED** [review]

File: examples/task_dynamic_foraging.py (hunk 1 above)
Comment (hanhou) [reply examples/task_dynamic_foraging.py]: removed!

**Review by jwong-nd — COMMENTED** [review]
Last Note!

Bruno wants the 'modifiable' fields in TaskParameters marked as modifiable. 
To do this, see here: 
https://github.com/AllenNeuralDynamics/aind-behavior-curriculum/blob/0c19397650fa71f27b202bf366a5b7235f5e4398/examples/task_creation.py#L20

This introduces an 'allow_modification' flag into the schema. 
See here: 
https://github.com/AllenNeuralDynamics/aind-behavior-curriculum/blob/0c19397650fa71f27b202bf366a5b7235f5e4398/examples/task_schema.json#L8

After that, everything is good to go

File: examples/task_dynamic_foraging.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    # Auto stop
+    MaxTrial: int = Field(..., title="Maximal number of trials", ge=0)
+    MaxTime: int = Field(..., title="Maximal session time (min)", ge=0)
+
+    RightValue_volume: float = Field(
+        3.0, title="Right reward size (uL)", ge=0.0
+    )
+    LeftValue_volume: float = Field(3.0, title="Left reward size (uL)", ge=0.0)
+
+    # Optional: Add additional validation to fields.
```
Comment (jwong-nd) [inline examples/task_dynamic_foraging.py:140]: Lines 140-145 can be removed as well

File: examples/task_dynamic_foraging.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+    ex_task = DynamicForagingTask(
+        version="1.0.0",
+        description=("Phase B in Han's slides "
+                     "(block = [10, 20, 5], p_sum = 0.8, p_ratio = [1:0])"),
+        task_parameters=ex_parameters,
+    )
+    print(ex_task)
+
+    import json
```
Comment (jwong-nd) [inline examples/task_dynamic_foraging.py:216]: Move 'import json' to the top of the file, this is not picked up by the linters

**Review by jwong-nd — CHANGES_REQUESTED** [review]
Add to comment above

File: examples/task_dynamic_foraging.py (hunk 2 above)
Comment (hanhou) [reply examples/task_dynamic_foraging.py]: done.

File: examples/task_dynamic_foraging.py (hunk 3 above)
Comment (hanhou) [reply examples/task_dynamic_foraging.py]: done.

**Review by jwong-nd — APPROVED** [review]

**Comment by hanhou** [comment]
> Last Note!
> 
> Bruno wants the 'modifiable' fields in TaskParameters marked as modifiable. To do this, see here:
> 
> https://github.com/AllenNeuralDynamics/aind-behavior-curriculum/blob/0c19397650fa71f27b202bf366a5b7235f5e4398/examples/task_creation.py#L20
> 
> This introduces an 'allow_modification' flag into the schema. See here:
> 
> https://github.com/AllenNeuralDynamics/aind-behavior-curriculum/blob/0c19397650fa71f27b202bf366a5b7235f5e4398/examples/task_schema.json#L8
> 


I'm a bit confused. Is `allow_modification` a custom field just for our own reference and has no effects on the schema? I don't see any difference in terms of validating `field_1 / 2` and `field_3 / 4` here, i.e., all of them are modifiable. 

```python
field_1: int = abc.ModifiableAttr(default=0, ge=0.0)
field_2: int = abc.ModifiableAttr(default=0, ge=0.0)
field_3: float = Field(default=0.5, ge=0.0, le=1.0)
field_4: float = Field(default=0.5, ge=0.0, le=1.0)
field_5: Literal["Immutable Field"] = "Immutable Field"
``` 

 Why don't we just use `Literal` versus the native `Field`?

**Comment by bruno-f-cruz** [comment]
> > Last Note!
> 
> > 
> 
> > Bruno wants the 'modifiable' fields in TaskParameters marked as modifiable. To do this, see here:
> 
> > 
> 
> > https://github.com/AllenNeuralDynamics/aind-behavior-curriculum/blob/0c19397650fa71f27b202bf366a5b7235f5e4398/examples/task_creation.py#L20
> 
> > 
> 
> > This introduces an 'allow_modification' flag into the schema. See here:
> 
> > 
> 
> > https://github.com/AllenNeuralDynamics/aind-behavior-curriculum/blob/0c19397650fa71f27b202bf366a5b7235f5e4398/examples/task_schema.json#L8
> 
> > 
> 
> 
> 
> 
> 
> I'm a bit confused. Is `allow_modification` a custom field just for our own reference and has no effects on the schema validation? Why don't we just use `Literal` versus the native `Field` to indicate immutable vs mutable fields?
> 
> 

Sorry missed this. As I mentioned in #8, I am not against using the const. That being said I thought Han was wanted to use this tag for something different. A const tag means that the schema will not allow modifications (either via the curriculum or via manual intervention) in other words, once const the value cannot be updated by the stage. I thought the modifiable tag was something to tag variables they could be modifiable outside the curriculum and still have the training be considered on curriculum. Maybe I am just being picky, but in my mind there is a clear difference between defining a const in the schema and allowing manual modification outside the curriculum.

**Comment by bruno-f-cruz** [comment]
Cleaning up hanging PR. Closing this as it is very out of date. Lets make a new PR with examples in the upcoming demo session.

**Pull request closed without merging**